src/utility/svd.jl

```diff
@@ -0,0 +1,207 @@
+import TensorKit:
+    SectorDict,
+    _empty_svdtensors,
+    _compute_svddata!,
+    _truncate!,
+    _implement_svdtruncation!,
+    _create_svdtensors
+
+# Plain copy of tsvd!(...) from TensorKit to lift alg type restriction
+function _tensorkit_svd!(
+    t::TensorMap;
+    trunc::TruncationScheme=TensorKit.NoTruncation(),
+    p::Real=2,
+    alg=TensorKit.SVD(),
+)
+    #early return
+    if isempty(blocksectors(t))
+        truncerr = zero(real(scalartype(t)))
+        return _empty_svdtensors(t)..., truncerr
+    end
+
+    S = spacetype(t)
+    Udata, Σdata, Vdata, dims = _compute_svddata!(t, alg)
+    if !isa(trunc, TensorKit.NoTruncation)
+        Σdata, truncerr = _truncate!(Σdata, trunc, p)
+        Udata, Σdata, Vdata, dims = _implement_svdtruncation!(t, Udata, Σdata, Vdata, dims)
+        W = S(dims)
+    else
+        truncerr = abs(zero(scalartype(t)))
+        W = S(dims)
+        if length(domain(t)) == 1 && domain(t)[1] ≅ W
+            W = domain(t)[1]
+        elseif length(codomain(t)) == 1 && codomain(t)[1] ≅ W
+            W = codomain(t)[1]
+        end
+    end
+    return _create_svdtensors(t, Udata, Σdata, Vdata, W)..., truncerr
+end
+
+# Wrapper struct around TensorKit's SVD algorithms
+@kwdef struct FullSVD
+    alg::Union{<:TensorKit.SVD,<:TensorKit.SDD} = TensorKit.SVD()
+    lorentz_broad::Float64 = 0.0
+end
+
+function svdwrap(t::AbstractTensorMap, alg::FullSVD; trunc=notrunc(), kwargs...)
+    # TODO: Replace _tensorkit_svd! with just tsvd eventually to use the full TensorKit machinery
+    return _tensorkit_svd!(copy(t); trunc, alg.alg)
+end
+
+function ChainRulesCore.rrule(
+    ::typeof(svdwrap), t::AbstractTensorMap, alg::FullSVD; trunc=notrunc(), kwargs...
+)
+    tsvd_return, tsvd!_pullback = ChainRulesCore.rrule(tsvd!, t; alg=TensorKit.SVD(), trunc)
+    function svdwrap_fullsvd_pullback(Δ)
+        return tsvd!_pullback(Δ)..., NoTangent()
+    end
+    return tsvd_return, svdwrap_fullsvd_pullback
+end
+
+# Wrapper around Krylov Kit's GKL iterative SVD solver
+@kwdef struct IterSVD
+    alg::KrylovKit.GKL = KrylovKit.GKL(; tol=1e-14, krylovdim=25)
+    howmany::Int = 20
+    lorentz_broad::Float64 = 0.0
+    alg_rrule::Union{GMRES,BiCGStab,Arnoldi} = GMRES(; tol=1e-14)
```

Comment:
I would keep this `alg_rrule` separate, maybe re-using the `hook_pullback` interface, to decouple the implementation of the forward and backward rules

Reply:
But in terms of usability, I think the SVD algorithm struct would be the best way to set the SVD adjoint parameters. So if I were to use `hook_pullback` in this case, I would need to store the `alg_rrule` and `lorentz_broad` settings in the `CTMRG` struct, which I also don't like that much.

Reply:
Ah I see, that's a very good point.

How about a wrapper then?
```julia
struct SVDrrule{F,T,R,E<:Real}
    svd_alg::F
    truncation_alg::T
    rrule_alg::R
    lorentz_broadening::E
end
```

In this case, we still at least semantically separate the rrule from the forward pass (which I really like), while also grouping them together, which I also quite like. It also looks to me to be quite flexible.

I think we should probably investigate (not just here, but throughout PEPSKit), if we want to include the truncation algorithm here or as a separate field. My guess would be that it's actually not a bad idea to keep them separate, considering we presumably might want to have varying `truncspace` settings throughout the unit cell, without changing the SVD algorithm.

Reply:
I have now added such a wrapper. It leads to the cool thing that you can easily choose the rrule with the corresponding algorithm struct and this decouples the forward SVD algorithm and its reverse rule. I decided to hide the `lorentz_broadening` away inside the `rrule_alg` since in general Lorentzian broadening might not be needed in every SVD implementation.

The only thing I'm unsure about is the naming; maybe something like `SVDWithRrule`, `SVDWithAdjoint`, `DiffSVD` or `DifferentiableSVD` would be more descriptive?